### PR TITLE
av.h - better document newAV_alloc_x/newAV_alloc_xz

### PR DESCRIPTION
**Pull request by richardleach**
PR Note: This is a revised version of #19006. The text is as previously proposed, other than I've tried to take into account @iabyn's concern that detailing exactly how array elements are initialized might unnecessarily complicate changes in the future.

### CODE REVIEW

File: av.h [hunk 1]
```diff
@@ -98,32 +98,46 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
 
-Note that the actual size requested is allocated. This is unlike
-av_extend(), which takes the maximum desired array index (AvMAX) as its
-"key" parameter, and enforces a minimum value for that of 3.
+newAV_alloc_x does not initialize the array elements - and so the
+expectation is that all should be initialized elsewhere prior to any
+potentials reads. newAV_alloc_xz does initialize the array elements.
 
-In other words, the following examples all result in an array that can
+As an example, the following examples all result in an array that can
```
Comment (khwilliamson) [inline av.h]: example occurring twice so close together reads awkwardly

File: av.h [hunk 2]
```diff
@@ -98,32 +98,46 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
+might be called directly:
+    av_extend(av, key);
 
-Similar to newAV(), but a SV* array is also allocated.
+Or it might be called implicitly when the first element is stored:
+    (void)av_store(av, 0, sv);
 
-This is similar to but more efficient than doing:
+Whether or not any as-yet-untouched array elements are initialized by
+av_extend depends upon whether the array is "real" at the time.
```
Comment (khwilliamson) [inline av.h]: I have a vague idea of what "real" is, but have never had to deal with it.  I think most readers will have even less of an idea.  Hopefully there is a discussion of this elsewhere that you can link to

**Review by khwilliamson — COMMENTED** [review]
Otherwise LGTM

**Comment by jkeenan** [comment]
Marginal note:  It appears this branch will clear up the build-time warning reported in https://github.com/Perl/perl5/issues/19029.  I'll double-check once this is merged.

File: av.h (hunk 1 above)
Comment (richardleach) [reply av.h]: Yes, I'll remove the leading "As an example, " there. Thanks.

File: av.h (hunk 2 above)
Comment (richardleach) [reply av.h]: Hmm, I've not really seen such a discussion and so only have my own piecemeal picture. Perhaps it would be better for this PR to more vaguely state something like:

"Unused array elements are typically initialized by av_extend, but this does not occur for some special arrays."

Even if that might just leave the reader to wonder what's special about them...

Comment (richardleach) [reply av.h]: @khwilliamson - I've reworded that sentence to avoid specifics. What do you think?

File: av.h [hunk 3]
```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
-Perl equivalent: C<my @array;>.
+The Perl equivalent is approximately C<my @array;>.
 
-=cut
-*/
+newAV does not allocate a SV* array.
+    AV *av = newAV();
 
-#define newAV()	MUTABLE_AV(newSV_type(SVt_PVAV))
+This is very useful when an AV is required, but population of it may be
```
Comment (khwilliamson) [inline av.h]: I think "but populating it' is better phrasing

File: av.h [hunk 4]
```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
+deferred, or even never actually take place. (Memory is not allocated
+unnecessarily.)
 
-/*
-=for apidoc newAV_alloc_x
-=for apidoc newAV_alloc_xz
+Subsequent SV* array allocation would be performed via av_extend. This
```
Comment (khwilliamson) [inline av.h]: Need to link to first mentions of other operations, and e,g, for subsequent mentions use C<av_extend>

File: av.h [hunk 5]
```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
+might be called directly:
+    av_extend(av, key);
 
-Similar to newAV(), but a SV* array is also allocated.
+Or it might be called implicitly when the first element is stored:
+    (void)av_store(av, 0, sv);
 
-This is similar to but more efficient than doing:
+Unused array elements are typically initialized by av_extend, but this
+is undesirable and will not be the case for some specific arrays.
```
Comment (khwilliamson) [inline av.h]: Where do they go to find the rules about which specific arrays and what

File: av.h [hunk 6]
```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
...
+
+For that scenario, newAV_alloc_x and newAV_alloc_xz can be used to create
+an AV and allocate a SV* array to fit the specified number of elements.
+(As a result, these macros MUST NOT be called with a size less than 1.)
 
-Note that the actual size requested is allocated. This is unlike
-av_extend(), which takes the maximum desired array index (AvMAX) as its
-"key" parameter, and enforces a minimum value for that of 3.
+newAV_alloc_x does not initialize the array elements - and so the
+expectation is that all should be initialized elsewhere prior to any
```
Comment (khwilliamson) [inline av.h]: s/should/will/
I think

File: av.h [hunk 7]
```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
 
 /*
 =for apidoc newAV
+=for apidoc_item newAV_alloc_x
+=for apidoc_item newAV_alloc_xz
 
-Creates a new AV.  The reference count is set to 1.
+These all create a new AV, setting the reference count to 1. They differ
+in when and how an accompanying SV* array is allocated and initialized.
```
Comment (khwilliamson) [inline av.h]: how about "They differ in allocating and populating the array of SV*s that always accompanies a non-empty AV."
or something similar

File: av.h (hunk 3 above)
Comment (richardleach) [reply av.h]: Ok, happy to change that.

File: av.h (hunk 4 above)
Comment (richardleach) [reply av.h]: Ok. Do the indented mentions need markup though, or are they fine bare?

File: av.h (hunk 5 above)
Comment (richardleach) [reply av.h]: Those specific arrays will be the non-`AvREAL()` ones, which don't seem to be described anywhere. I'm happy to have a go at documenting "real/unreal" arrays (in _perlguts_ or in code comments?)  but would rather not do as part of this PR.

For now, should it just say:
`Unused array elements are typically initialized by av_extend.`?

File: av.h (hunk 6 above)
Comment (richardleach) [reply av.h]: Works for me.

File: av.h (hunk 7 above)
Comment (richardleach) [reply av.h]: I've changed it to something similar.

File: av.h (hunk 5 above)
Comment (richardleach) [reply av.h]: Or we hold this PR until AvREAL is documented, and then go back to the original text of:
```
Whether or not any as-yet-untouched array elements are initialized by
av_extend depends upon whether the array is "real" at the time.
```

Comment (richardleach) [reply av.h]: I take that back, there are some comments at the top of `av.h`. I'll have a look at converting them to apidoc and expanding them.

Comment (richardleach) [reply av.h]: I've converted this sentence to:
```
Unused array elements are typically initialized by C<av_extend>. (Only
core maintainers should have need to concern themseleves with when that
is not the case. Refer to F<av.h> and F<av.c> for the differences between 
real and fake AVs.)
```
And opened https://github.com/Perl/perl5/pull/19062 as a separate concern for the documentation of real/fake AVs.
